include/stl2/detail/range/primitives.hpp

```diff
@@ -42,25 +79,54 @@ STL2_OPEN_NAMESPACE {
 		)
 	}
 
-	Range{R}
-	constexpr iter_difference_t<iterator_t<R>> distance(R&& r)
-	noexcept(noexcept(__stl2::distance(__stl2::begin(r), __stl2::end(r))))
-	{
-		const iter_difference_t<iterator_t<R>> n =
-			__stl2::distance(__stl2::begin(r), __stl2::end(r));
-		STL2_EXPECT(n >= 0);
-		return n;
-	}
+	struct __distance_fn {
+		template<Iterator I, Sentinel<I> S>
+		constexpr iter_difference_t<I> operator()(I first, S last) const
+		// [[expects axiom: reachable(first, last)]]
+		STL2_NOEXCEPT_RETURN(
+			ext::enumerate(std::move(first), std::move(last)).first
+		)
 
-	SizedRange{R}
-	constexpr iter_difference_t<iterator_t<R>> distance(R&& r)
-	noexcept(noexcept(__stl2::size(r)))
-	{
-		const auto n =
-			static_cast<iter_difference_t<iterator_t<R>>>(__stl2::size(r));
-		STL2_EXPECT(n >= 0);
-		return n;
-	}
+		template<Iterator I, SizedSentinel<I> S>
+		constexpr iter_difference_t<I> operator()(I first, S last) const
+		noexcept(noexcept(last - first))
+		// [[expects axiom: reachable(first, last)]]
+		{
+			auto d = last - first;
+			STL2_EXPECT(d >= 0);
+			return d;
+		}
+
+		template<Iterator I>
+		requires SizedSentinel<I, I>
+		constexpr iter_difference_t<I> operator()(I first, I last) const
+		// [[expects axiom: reachable(first, last)]]
+		STL2_NOEXCEPT_RETURN(
+			last - first
+		)
+
+		template<Range R>
+		constexpr iter_difference_t<iterator_t<R>> operator()(R&& r) const
+		noexcept(noexcept(std::declval<__distance_fn const&>()(__stl2::begin(r), __stl2::end(r))))
```

Comment:
Boo GCC for not fixing https://gcc.gnu.org/bugzilla/show_bug.cgi?id=52869. (No change requested.)